File: 238-product-of-array-except-self/238-product-of-array-except-self.scala
object Solution {
    def productExceptSelf(nums: Array[Int]): Array[Int] = {
        val res = Array.fill(nums.length)(0)
        
        var cur = 1
        for(i <- 0 to nums.length - 1) {
            res(i) = cur
            cur *= nums(i)
        }
        
        cur = 1
        for(i <- (0 to nums.length - 1).reverse) {
            res(i) *= cur
            cur *= nums(i)
        }
        
        res
    }
}

File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        
        Arrays.fill(res, 1);
        
        for(int i = 0, cur = 1; i < nums.length; i++) {
            res[i] *= cur;
            cur *= nums[i];
        }
        
        for(int i = 0, cur = 1; i < nums.length; i++) {
            res[nums.length - 1 - i] *= cur;
            cur *= nums[nums.length - 1 - i];
        }
        
        return res;
    }
}